hw1/bashcommands.py

```diff
@@ -0,0 +1,141 @@
+        current target argument index in the list of all target arguments provided to the result function,
+        total number of target arguments provided to the result function,
+        accumulator of type R which is equal to the return value of the f function on the previous iteration
+        :return:
+        """
+        accumulator = initial_value
+
+        @wraps(f)
+        def f_for_many(*many_args):
+            nonlocal accumulator
+            for arg_index, arg in enumerate(many_args):
+                accumulator = f(arg, arg_index, len(many_args), accumulator)
+            return accumulator
+
+        return f_for_many
+
+    return decorator
+
+
+def print_line(line: str) -> None:
+    """
+    This is needed to fight special cases when lines end not with a newline character but with EOF or some weird
+    character. Supports only \n newlines for simplicity.
+    """
+    print(line if len(line) == 0 or line[-1] == "\n" else f"{line}\n", end="")
+
+
+@for_each_argument(initial_value=(0, 0, 0))
+def wc(
+    filepath: str, cur_filepath_index: int, n_filepaths: int, total_counters: Tuple[int, int, int]
+) -> Tuple[int, int, int]:
+    total_lines, total_words, total_bytes = total_counters
+    cur_lines, cur_words, cur_bytes = 0, 0, 0
+    try:
+        with open(filepath) as f:
+            for line in f:
+                cur_lines += 1
+                cur_words += len(line.split())
+                cur_bytes += len(bytes(line, encoding="utf8"))
+        total_lines += cur_lines
+        total_words += cur_words
+        total_bytes += cur_bytes
+        print(f"lines : {cur_lines}, words : {cur_words}, bytes : {cur_bytes} for {filepath}")
+    except FileNotFoundError:
+        print(filepath, "no such file")
+
+    is_last_file = cur_filepath_index == n_filepaths - 1
+    if is_last_file:
+        print(f"lines : {total_lines}, words : {total_words}, bytes : {total_bytes} in total\n")
+    return total_lines, total_words, total_bytes
+
+
+def is_blank(string: str) -> bool:
+    blank_chars = {"\n", "\r", "\r\n", "\t"}
+    return set(string) | blank_chars == blank_chars
+
+
+@for_each_argument(initial_value=1)
+def nl(filepath: str, _, __, line_count: int) -> int:
+    cur_line_count = line_count
+    try:
+        with open(filepath) as f:
+            for line in f:
+                if is_blank(line):
+                    print()
+                else:
+                    print_line(f"{cur_line_count}\t{line}")
+                    cur_line_count += 1
+    except FileNotFoundError:
+        print(filepath, "no such file")
+
+    return cur_line_count
+
+
+@for_each_argument()
+def head(filepath: str, _, n_files: int, __) -> None:
+    if n_files > 1:
+        print(f"==> {filepath} <==")
+    try:
+        with open(filepath) as f:
+            for line_num, line in enumerate(f):
+                if line_num >= 10:
+                    break
+                print_line(line)
```

Comment:
Let's write as `while`, because `for` is usually used when you actually want to run the whole cycle. Here, you do a typically `while` thing.